### PR TITLE
Refractor properties and features for dumping

### PR DESCRIPTION
**Pull request by alejoe91**
Final dumps! I think this is super important to save the status of recording and sortings (e.g. after extracting waveforms, pca, etc...)

- Added the consept of `key_properties` that for RecordingExtractors are `group` and `location`

- This cleans up property and feature handling, now done at the base class (`self._properties` and `self._features`).

- `make_serialized_dict()` allows saves `key_properties` by default

- 3 dump functions: `dump_to_dict`, `dump_to_json` (both save key properties) and `dump_to_pickle`, that enable saving all features and properties

- The `load_extractor_from_dict` and `load_extractor_from_json` have been modified to load the properties and features and the `load_extractor_from_pickle` has been added

### CODE REVIEW

File: spikeextractors/baseextractor.py [hunk 1]
```diff
@@ -32,47 +39,47 @@ def make_serialized_dict(self):
...
-                    dump_dict['groups'] = groups
         else:
             dump_dict = {'class': class_name, 'module': module, 'kwargs': {},
                          'version': imported_module.__version__, 'dumpable': False}
-
+        dump_dict = self._include_properties_and_features_in_dict(dump_dict, include_properties, include_features)
         return dump_dict
 
-    def dump(self, file_name=None, folder_path=None):
+    def dump(self, file_path=None, dump_properties_and_features=True):
```
Comment (samuelgarcia) [inline spikeextractors/baseextractor.py:48]: I think I would put the same signautre than ```make_serialized_dict``` with include_properties=None, include_features=None.
But by default I would put include_properties='all' and 'all would be all feature or properties.

**Comment by samuelgarcia** [comment]
I don't known what to think.
This dump is not as elegant as the previous one with only json.

I would propose a dump_json (as before) and a dump that would use only pickle include for the dict.
This hybrid json/pickle don't make me so happy.
Even I understand the need.

**Comment by alejoe91** [comment]
I know, but we need to dump data as well...I don't want to compute waveforms more than one time and we also make analysis that we can save as property/fetaure and reload later. Any other suggestion?

**Comment by samuelgarcia** [comment]
My suggestion would be:
  * keep of course the serilized to dict.
  * dump use pickle and only pickle for dict and properties but in the same file.
  * a separate dump_json with no properties.

It would be more coherent no ?

**Comment by alejoe91** [comment]
> My suggestion would be:
> 
> * keep of course the serilized to dict.
> * dump use pickle and only pickle for dict and properties but in the same file.
> * a separate dump_json with no properties.
> 
> It would be more coherent no ?

So then one would need to call 2 functions to save and 2 functions to load the data? Or you suggest to dump 1) dict, 2) properties, 2) features in the same pickle?

**Comment by alejoe91** [comment]
so then the dump would save a pickle and the dump_json would save a json..that can work

**Comment by samuelgarcia** [comment]
I think that **_key_properties** is not the appropriate naming.

**Comment by alejoe91** [comment]
Apart from that, I think this is cool :)

spiketoolkit is ready as well

**Comment by alejoe91** [comment]
@samuelgarcia this is ready to be merged

**Comment by samuelgarcia** [comment]
I think the week end break my chance to review this.
Have you propagate to spikecomparison.load_probe_file_inplace ?
This is quite important for GTstudy.

**Comment by alejoe91** [comment]
Ahhaah hi Sam! The API is the same and all tests from other packages are fine :) you can review it now and I can make extra changes, but this is exactly what we agreed on! (Except the naming)

**Pull request merged**